addons/hw_scale/controllers/main.py

```diff
@@ -338,23 +343,30 @@ def run(self):
         self.device = None
 
         while True:
-            if self.device:
-                old_weight = self.weight
-                self.read_weight()
-                if self.weight != old_weight:
-                    _logger.info('New Weight: %s, sleeping %ss', self.weight, self.protocol.newWeightDelay)
-                    time.sleep(self.protocol.newWeightDelay)
-                    if self.weight and self.protocol.autoResetWeight:
-                        self.weight = 0
+            if not self.disabled:
+                if self.device:
+                    old_weight = self.weight
+                    self.read_weight()
+                    if self.weight != old_weight:
+                        _logger.info('New Weight: %s, sleeping %ss', self.weight, self.protocol.newWeightDelay)
+                        time.sleep(self.protocol.newWeightDelay)
+                        if self.weight and self.protocol.autoResetWeight:
+                            self.weight = 0
+                    else:
+                        _logger.info('Weight: %s, sleeping %ss', self.weight, self.protocol.weightDelay)
+                        time.sleep(self.protocol.weightDelay)
+                        self.disabled = True
                 else:
-                    _logger.info('Weight: %s, sleeping %ss', self.weight, self.protocol.weightDelay)
-                    time.sleep(self.protocol.weightDelay)
+                    with self.scalelock:
+                        self.device = self.get_device()
+                    if not self.device:
+                        # retry later to support "plug and play"
+                        time.sleep(10)
+                    else: 
+                        self.disabled = self.protocol.disable
             else:
-                with self.scalelock:
-                    self.device = self.get_device()
-                if not self.device:
-                    # retry later to support "plug and play"
-                    time.sleep(10)
+                time.sleep(10)
```

Comment:
This long sleep gives a slow experience to users - whenever they enter the POS weighing screen they have to wait up to 10 seconds until the scale picks up the weight. I have received feedback where they tell me that it's faster for them to type the weight manually in the POS now.
Either set this sleep to a very small value, e.g. 0.1 sec, or use another technique to exit the sleep faster when a call arrives. One way to do it is to use a `Threading.Event`. Setup an Event in the scale_thread, and instead of `time.sleep()`, use `self.event.wait(10)`, and wake it up instantly from any read call using `self.event.set()`. 
Just be sure to `clear()` the event after waking up, so the next `wait()` works again.

Reply:
setting the sleep to a small value will make the scale beep all the time again, this is not a possibility. Also waiting for an event to come in is not an option since we have to send a weight request to the scale to get the new value.

@pha-odoo did you find a way to improve this when rewriting to IoT?

Reply:
> setting the sleep to a small value will make the scale beep all the time again, this is not a possibility. 

Could you clarify this? As far as I can remember the beep comes from reading the weight. This particular sleep is what happens when the reading thread is `disabled`, hence no reading is taking place. How will this make the scale beep?

> Also waiting for an event to come in is not an option since we have to send a weight request to the scale to get the new value.

How is that a problem? The `/hw_proxy/scale_read/` method will call `get_weight` which will set() the event to instantly release the sleeping thread, pretty much similar to what it does with the `disabled` flag now. And when the thread goes back to sleeping because it's disabled again, it will clear() the event in preparation for being woken again.

Reply:
@odony I misunderstood what you where referring to at first. Thanks for the help, I will look into it and keep you posted.